File: modules/media-tools/ios/MediaToolsModule.swift
import AVFoundation
import CoreLocation
import ExpoModulesCore
import Photos

/**
 * iOS side of `media-tools` (§7, §8): source video properties, batched asset sizes, and a
 * metadata-carrying gallery save. The TypeScript surface is identical to Android's; callers branch
 * on `getCapabilities`, never on the platform.
 *
 * Photos keeps capture date and location in its own database — unlike Android's MediaStore it
 * never re-derives them from the file — so a save sets them in the creation request and they
 * stick. There is no MP4 atom stamping here for the same reason.
 *
 * The service functions are no-ops by design: iOS has no foreground service, and §7's background
 * continuation rides the compressor's own background task instead (see `core/background`).
 */
public class MediaToolsModule: Module {
  public func definition() -> ModuleDefinition {
    Name("MediaTools")

    Function("getCapabilities") { () -> [String: Bool] in
      [
        "videoProperties": true,
        "assetSizes": true,
        "captureDateWriteBack": true,
        "locationWriteBack": true,
        "foregroundService": false,
        "librarySave": true
      ]
    }

    AsyncFunction("saveVideo") { (options: SaveVideoInput, promise: Promise) in
      saveVideo(options, promise: promise)
    }

    AsyncFunction("readVideoProperties") { (assetId: String, promise: Promise) in
      readVideoProperties(assetId: assetId, promise: promise)
    }

    AsyncFunction("readAssetSizes") { (assetIds: [String]) -> [String: Double] in
      readAssetSizes(assetIds)
    }

    AsyncFunction("applyAssetMetadata") { (assetId: String, metadata: AssetMetadataInput, promise: Promise) in
      applyAssetMetadata(assetId: assetId, metadata: metadata, promise: promise)
    }

    AsyncFunction("startCompressionService") { (_: ServiceNotification) in }

    AsyncFunction("updateCompressionProgress") { (_: ServiceNotification) in }

    AsyncFunction("stopCompressionService") { }
  }
}

// MARK: - Asset identity

/**
 * The JS layer's asset ids are the `ph://…` URIs expo-media-library hands out (they double as
 * image and player sources). Photos wants the bare local identifier; every function converts on
 * the way in and — crucially — keys results by the exact id it was given, because the size
 * index's map lookups depend on the round trip being identity.
 */
private let phScheme = "ph://"

private func localIdentifier(fromAssetId assetId: String) -> String {
  assetId.hasPrefix(phScheme) ? String(assetId.dropFirst(phScheme.count)) : assetId
}

private func assetId(fromLocalIdentifier identifier: String) -> String {
  phScheme + identifier
}

private func fetchAsset(_ assetId: String) -> PHAsset? {
  PHAsset.fetchAssets(
    withLocalIdentifiers: [localIdentifier(fromAssetId: assetId)],
    options: nil
  ).firstObject
}

// MARK: - Sizes

private func readAssetSizes(_ assetIds: [String]) -> [String: Double] {
  if assetIds.isEmpty { return [:] }

  var inputIds: [String: String] = [:]
  for id in assetIds {
    inputIds[localIdentifier(fromAssetId: id)] = id
  }

  var sizes: [String: Double] = [:]
  let fetched = PHAsset.fetchAssets(withLocalIdentifiers: Array(inputIds.keys), options: nil)
  fetched.enumerateObjects { asset, _, _ in
    guard let inputId = inputIds[asset.localIdentifier] else { return }
    if let size = fileSize(of: asset), size > 0 { sizes[inputId] = size }
  }

  return sizes
}

/**
 * The bytes behind the asset, read from resource metadata alone — no file I/O and no iCloud
 * download, which is what lets the §4 size index sweep a 5,000-video library at background
 * priority. The edited rendition is preferred over the original recording because it is what the
 * user's gallery shows and what a compression would read.
 *
 * `fileSize` is a key the resource answers but the SDK never formalized; the `responds` guard
 * turns a future removal into "size unknown" (sorted last, excluded from filters) instead of an
 * uncatchable KVC exception.
 */
private func fileSize(of asset: PHAsset) -> Double? {
  let resources = PHAssetResource.assetResources(for: asset)
  let resource =
    resources.first { $0.type == .fullSizeVideo }
    ?? resources.first { $0.type == .video }
    ?? resources.first
  guard let resource, resource.responds(to: NSSelectorFromString("fileSize")) else { return nil }
  return (resource.value(forKey: "fileSize") as? NSNumber)?.doubleValue
}

// MARK: - Source properties

/**
 * The capture facts come straight off the `PHAsset`; the track facts (frame rate, rotation,
 * bitrate) need the `AVAsset` — requested WITHOUT network access on purpose. On the Selected
 * screen this keeps an iCloud-offloaded video from downloading just to show its details; on the
 * compression path `resolveLocalPath` has already pulled the bytes down before this runs, so the
 * request is served locally. When the AVAsset genuinely is not available, the track fields read
 * null and the caller falls back to its assumed frame rate.
 */
private func readVideoProperties(assetId: String, promise: Promise) {
  guard let asset = fetchAsset(assetId) else {
    // A stale or deleted asset simply has no properties; the caller falls back.
    promise.resolve(nil)
    return
  }

  let options = PHVideoRequestOptions()
  options.isNetworkAccessAllowed = false
  options.version = .current

  PHImageManager.default().requestAVAsset(forVideo: asset, options: options) { avAsset, _, _ in
    let track = avAsset?.tracks(withMediaType: .video).first

    promise.resolve([
      // iOS albums are not directories: there is no folder to report, and none to save into.
      "folder": NSNull(),
      "sizeBytes": orNull(fileSize(of: asset)),
      "frameRate": orNull(positive(track.map { Double($0.nominalFrameRate) })),
      "rotationDegrees": track.map(rotationDegrees(of:)) ?? 0,
      "bitrate": orNull(positive(track.map { Double($0.estimatedDataRate) }).map { $0.rounded() }),
      "capturedAtMs": orNull(asset.creationDate.map(epochMs(of:))),
      "location": orNull(asset.location.map {
        ["latitude": $0.coordinate.latitude, "longitude": $0.coordinate.longitude]
      })
    ] as [String: Any])
  }
}

/** The track's display rotation, folded to the 0/90/180/270 the pipeline works in. */
private func rotationDegrees(of track: AVAssetTrack) -> Int {
  let transform = track.preferredTransform
  let radians = atan2(Double(transform.b), Double(transform.a))
  let degrees = Int((radians * 180 / .pi).rounded())
  return ((degrees % 360) + 360) % 360
}

// MARK: - Saving to the library

/**
 * §8: saves a finished encode as a new gallery asset carrying the source's capture date and
 * location, both set in the same change block the asset is created in. What could not be carried
 * is reported, not faked: iOS has no settable modification date (`PHAsset.modificationDate` is
 * system-owned) and no folders.
 */
private func saveVideo(_ options: SaveVideoInput, promise: Promise) {
  let url = fileUrl(fromPath: options.path)
  guard FileManager.default.fileExists(atPath: url.path) else {
    promise.reject(SaveFailedException("there is no file at \(url.path)"))
    return
  }

  var placeholderIdentifier: String?

  PHPhotoLibrary.shared().performChanges({
    let creation = PHAssetCreationRequest.forAsset()

    let resourceOptions = PHAssetResourceCreationOptions()
    resourceOptions.originalFilename = options.filename
    // The workspace copy is discarded right after a successful save; moving instead of copying
    // skips holding the video on disk twice.
    resourceOptions.shouldMoveFile = true
    creation.addResource(with: .video, fileURL: url, options: resourceOptions)

    if let capturedAtMs = options.capturedAtMs {
      creation.creationDate = date(fromEpochMs: capturedAtMs)
    }
    if let location = location(latitude: options.latitude, longitude: options.longitude) {
      creation.location = location
    }

    placeholderIdentifier = creation.placeholderForCreatedAsset?.localIdentifier
  }, completionHandler: { success, error in
    guard success, let identifier = placeholderIdentifier else {
      promise.reject(SaveFailedException(
        error?.localizedDescription ?? "the photo library refused the save"
      ))
      return
    }

    promise.resolve([
      "assetId": assetId(fromLocalIdentifier: identifier),
      "report": saveReport(for: options)
    ] as [String: Any])
  })
}

/**
 * What the creation request carried. The change block is atomic — reaching here means Photos
 * accepted every value in it — so the report is built from what was requested, with the fields
 * iOS can never store listed as skipped with the reason.
 */
private func saveReport(for options: SaveVideoInput) -> [String: Any] {
  var applied: [String] = []
  var skipped: [[String: String]] = []

  if options.capturedAtMs != nil { applied.append("capturedAt") }
  if location(latitude: options.latitude, longitude: options.longitude) != nil {
    applied.append("location")
  }
  if options.modifiedAtMs != nil {
    skipped.append(["field": "modifiedAt", "reason": noSettableModifiedDate])
  }

  return ["applied": applied, "skipped": skipped]
}

// MARK: - Metadata write-back

/**
 * §8's fallback path for an asset that already exists: the same fields, written through
 * `PHAssetChangeRequest` and then verified by refetching the asset — a change Photos ignored
 * looks exactly like one it accepted, and the report promises what a gallery will actually show.
 */
private func applyAssetMetadata(assetId: String, metadata: AssetMetadataInput, promise: Promise) {
  guard let asset = fetchAsset(assetId) else {
    promise.resolve(writeBackReport(assetId: assetId, metadata: metadata, failure: "the asset no longer exists"))
    return
  }

  let capturedAt = metadata.capturedAtMs.map(date(fromEpochMs:))
  let newLocation = location(latitude: metadata.latitude, longitude: metadata.longitude)

  PHPhotoLibrary.shared().performChanges({
    let change = PHAssetChangeRequest(for: asset)
    if let capturedAt { change.creationDate = capturedAt }
    if let newLocation { change.location = newLocation }
  }, completionHandler: { success, error in
    promise.resolve(writeBackReport(
      assetId: assetId,
      metadata: metadata,
      failure: success ? nil : (error?.localizedDescription ?? "the photo library refused the change")
    ))
  })
}

private func writeBackReport(
  assetId: String,
  metadata: AssetMetadataInput,
  failure: String?
) -> [String: Any] {
  var applied: [String] = []
  var skipped: [[String: String]] = []
  // One refetch answers every verification below; nil means nothing can have been written.
  let saved = failure == nil ? fetchAsset(assetId) : nil

  if let expectedMs = metadata.capturedAtMs {
    let actualMs = saved?.creationDate.map(epochMs(of:))
    // Photos stores sub-second dates; a written value reads back within rounding.
    if let actualMs, abs(actualMs - expectedMs) < 1000 {
      applied.append("capturedAt")
    } else {
      skipped.append(field("capturedAt", failure.map { "The photo library threw: \($0)" }
        ?? "Wrote \(Int(expectedMs)), read back \(actualMs.map { String(Int($0)) } ?? "nothing")."))
    }
  }

  if metadata.modifiedAtMs != nil {
    skipped.append(field("modifiedAt", noSettableModifiedDate))
  }

  if metadata.latitude != nil && metadata.longitude != nil {
    if saved?.location != nil {
      applied.append("location")
    } else {
      skipped.append(field("location", failure.map { "The photo library threw: \($0)" }
        ?? "The location did not read back from the saved asset."))
    }
  }

  return ["applied": applied, "skipped": skipped]
}

// MARK: - Shared helpers

private let noSettableModifiedDate =
  "iOS does not expose a settable modification date on photo library assets."

private func field(_ name: String, _ reason: String) -> [String: String] {
  ["field": name, "reason": reason]
}

private func epochMs(of date: Date) -> Double {
  date.timeIntervalSince1970 * 1000
}

private func date(fromEpochMs epochMs: Double) -> Date {
  Date(timeIntervalSince1970: epochMs / 1000)
}

private func location(latitude: Double?, longitude: Double?) -> CLLocation? {
  guard let latitude, let longitude else { return nil }
  return CLLocation(latitude: latitude, longitude: longitude)
}

private func fileUrl(fromPath path: String) -> URL {
  if let url = URL(string: path), url.scheme == "file" { return url }
  return URL(fileURLWithPath: path)
}

private func positive(_ value: Double?) -> Double? {
  guard let value, value > 0 else { return nil }
  return value
}

/** Absent facts cross the bridge as explicit nulls, matching the TypeScript nullable fields. */
private func orNull(_ value: Any?) -> Any {
  value ?? NSNull()
}

internal final class SaveFailedException: GenericException<String> {
  override var reason: String {
    "Saving to the photo library failed: \(param)."
  }
}

struct SaveVideoInput: Record {
  @Field var path: String = ""

  @Field var filename: String = "video.mp4"

  /** Android's MediaStore folder; iOS has no folders — albums are not directories — so unused. */
  @Field var folder: String?

  @Field var capturedAtMs: Double?

  @Field var modifiedAtMs: Double?

  @Field var latitude: Double?

  @Field var longitude: Double?
}

struct ServiceNotification: Record {
  @Field var title: String = "Compressing video"

  @Field var progress: Int = 0

  @Field var elapsed: String = ""

  @Field var remaining: String = ""
}

struct AssetMetadataInput: Record {
  @Field var capturedAtMs: Double?

  @Field var modifiedAtMs: Double?

  @Field var latitude: Double?

  @Field var longitude: Double?
}

File: modules/media-tools/android/src/main/java/expo/modules/mediatools/MediaToolsModule.kt
package expo.modules.mediatools

import android.content.ContentUris
import android.content.ContentValues
import android.content.Context
import android.content.Intent
import android.media.MediaMetadataRetriever
import android.net.Uri
import android.os.Build
import android.os.Handler
import android.os.Looper
import android.provider.MediaStore
import android.util.Log
import expo.modules.kotlin.exception.Exceptions
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.records.Field
import expo.modules.kotlin.records.Record
import java.io.File
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale

/**
 * The four things the app needs that no JavaScript dependency can provide (§7, §8):
 * source video properties, batched asset sizes, capture metadata write-back, and a foreground
 * service for background compression.
 *
 * The TypeScript surface is identical on both platforms; the iOS implementation reports its
 * capabilities as false until it is written, so callers branch on the result, never on the platform.
 */
class MediaToolsModule : Module() {
  private companion object {
    /** §5's output container; the encoder never produces anything else. */
    const val MP4_MIME_TYPE = "video/mp4"

    /** `adb logcat -s MediaTools` follows a save end to end. */
    const val TAG = "MediaTools"
  }

  private val context: Context
    get() = appContext.reactContext ?: throw Exceptions.ReactContextLost()

  private val resolver get() = context.contentResolver

  override fun definition() = ModuleDefinition {
    Name("MediaTools")

    Function("getCapabilities") {
      mapOf(
        "videoProperties" to true,
        "assetSizes" to true,
        "captureDateWriteBack" to true,
        // MediaStore's LATITUDE/LONGITUDE columns were removed in Android 10 and nothing here can
        // write an ISO-6709 atom into the MP4, so location genuinely cannot be carried over.
        "locationWriteBack" to false,
        "foregroundService" to true,
        "librarySave" to true
      )
    }

    AsyncFunction("saveVideo") { options: SaveVideoInput ->
      saveVideo(options)
    }

    AsyncFunction("readVideoProperties") { assetId: String ->
      readVideoProperties(assetId)
    }

    AsyncFunction("readAssetSizes") { assetIds: List<String> ->
      readAssetSizes(assetIds)
    }

    AsyncFunction("applyAssetMetadata") { assetId: String, metadata: AssetMetadataInput ->
      applyAssetMetadata(assetId, metadata)
    }

    AsyncFunction("startCompressionService") { options: ServiceNotification ->
      startService(options)
    }

    AsyncFunction("updateCompressionProgress") { options: ServiceNotification ->
      updateNotification(options)
    }

    AsyncFunction("stopCompressionService") {
      stopService()
    }
  }

  // MARK: - Source properties

  private fun readVideoProperties(assetId: String): Map<String, Any?>? {
    val uri = Uri.parse(assetId)
    val retriever = MediaMetadataRetriever()

    return try {
      retriever.setDataSource(context, uri)

      mapOf(
        "folder" to queryString(uri, MediaStore.MediaColumns.RELATIVE_PATH),
        "sizeBytes" to querySize(uri),
        // Parentheses required: infix `to` binds tighter than `?:`.
        "frameRate" to (
          retriever.extract(MediaMetadataRetriever.METADATA_KEY_CAPTURE_FRAMERATE)?.toFloatOrNull()
            ?: deriveFrameRate(retriever)
          ),
        "rotationDegrees" to
          (retriever.extract(MediaMetadataRetriever.METADATA_KEY_VIDEO_ROTATION)?.toIntOrNull() ?: 0),
        "bitrate" to retriever.extract(MediaMetadataRetriever.METADATA_KEY_BITRATE)?.toIntOrNull(),
        "capturedAtMs" to queryCapturedAt(uri),
        "location" to parseIso6709(
          retriever.extract(MediaMetadataRetriever.METADATA_KEY_LOCATION)
        )
      )
    } catch (error: Exception) {
      // A stale or unreadable asset simply has no properties; the caller falls back.
      null
    } finally {
      retriever.release()
    }
  }

  /** Frame count over duration — the capture-framerate key is absent on plenty of devices. */
  private fun deriveFrameRate(retriever: MediaMetadataRetriever): Float? {
    if (Build.VERSION.SDK_INT < Build.VERSION_CODES.P) return null

    val frames = retriever.extract(MediaMetadataRetriever.METADATA_KEY_VIDEO_FRAME_COUNT)
      ?.toIntOrNull() ?: return null
    val durationMs = retriever.extract(MediaMetadataRetriever.METADATA_KEY_DURATION)
      ?.toLongOrNull() ?: return null
    if (durationMs <= 0) return null

    return frames * 1000f / durationMs
  }

  private fun MediaMetadataRetriever.extract(key: Int): String? = extractMetadata(key)

  // MARK: - Sizes

  /** One cursor for the whole batch — the media store already indexes SIZE. */
  private fun readAssetSizes(assetIds: List<String>): Map<String, Double> {
    if (assetIds.isEmpty()) return emptyMap()

    val byRowId = assetIds.mapNotNull { id ->
      runCatching { ContentUris.parseId(Uri.parse(id)) }.getOrNull()?.let { it to id }
    }.toMap()
    if (byRowId.isEmpty()) return emptyMap()

    val placeholders = byRowId.keys.joinToString(",") { "?" }
    val sizes = mutableMapOf<String, Double>()

    context.contentResolver.query(
      MediaStore.Video.Media.EXTERNAL_CONTENT_URI,
      arrayOf(MediaStore.Video.Media._ID, MediaStore.Video.Media.SIZE),
      "${MediaStore.Video.Media._ID} IN ($placeholders)",
      byRowId.keys.map { it.toString() }.toTypedArray(),
      null
    )?.use { cursor ->
      val idColumn = cursor.getColumnIndexOrThrow(MediaStore.Video.Media._ID)
      val sizeColumn = cursor.getColumnIndexOrThrow(MediaStore.Video.Media.SIZE)

      while (cursor.moveToNext()) {
        val assetId = byRowId[cursor.getLong(idColumn)] ?: continue
        val size = cursor.getLong(sizeColumn)
        if (size > 0) sizes[assetId] = size.toDouble()
      }
    }

    return sizes
  }

  private fun querySize(uri: Uri): Double? = queryLong(uri, MediaStore.Video.Media.SIZE)?.toDouble()

  private fun queryCapturedAt(uri: Uri): Double? =
    queryLong(uri, MediaStore.Video.Media.DATE_TAKEN)?.takeIf { it > 0 }?.toDouble()

  private fun queryLong(uri: Uri, column: String): Long? =
    context.contentResolver.query(uri, arrayOf(column), null, null, null)?.use { cursor ->
      if (cursor.moveToFirst() && !cursor.isNull(0)) cursor.getLong(0) else null
    }

  private fun queryString(uri: Uri, column: String): String? =
    runCatching {
      context.contentResolver.query(uri, arrayOf(column), null, null, null)?.use { cursor ->
        if (cursor.moveToFirst() && !cursor.isNull(0)) cursor.getString(0) else null
      }
    }.getOrNull()

  // MARK: - Saving to the library

  /**
   * §8: saves a finished encode into the gallery, in the source's own folder, carrying its dates.
   *
   * The dates go into the MP4's own header atoms first, because the media store's columns are not
   * durable: publishing the row triggers an asynchronous scan that re-derives DATE_TAKEN from the
   * file's `creation_time` — which the encoder wrote as "now" — and can land after any column
   * write of ours. Once the file itself carries the source's dates, every scan converges on them.
   * The column writes at insert and publish then only cover the window before that scan runs.
   *
   * `IS_PENDING` keeps the row invisible to other apps until the bytes are in, so no gallery ever
   * shows a half-written video.
   */
  private fun saveVideo(options: SaveVideoInput): String {
    // Request and outcome bracket every save, so a metadata bug report needs no extra build.
    Log.i(TAG, "[save] request ${describe(options)}")

    stampOutputFileDates(options)

    val uri = insertPendingRow(options)
    copyBytesInto(uri, options.path)
    publish(uri, options)
    assertDates(uri, options)

    Log.i(TAG, "[save] final ${readColumns(uri)}")
    return uri.toString()
  }

  /**
   * Patches the encode's `mvhd`/`tkhd`/`mdhd` atoms to the source's dates before the bytes leave
   * our temp directory. The file is workspace-owned and about to be discarded, so mutating it is
   * safe — and a patch failure only means falling back to the column-write path.
   */
  private fun stampOutputFileDates(options: SaveVideoInput) {
    val creationMs = options.capturedAtMs?.toLong() ?: return
    val modificationMs = options.modifiedAtMs?.toLong() ?: creationMs
    val file = File(options.path.removePrefix("file://"))

    runCatching { Mp4Dates.apply(file, creationMs, modificationMs) }
      .onSuccess { atoms ->
        // Zero atoms means the dates rest on the column writes alone — the first thing to check
        // if a saved copy ever shows the wrong date again.
        if (atoms == 0) Log.w(TAG, "[save] no MP4 atoms patched; dates rely on column writes")
        else Log.i(TAG, "[save] stamped $atoms MP4 atoms with the source dates")
      }
      .onFailure { Log.w(TAG, "[save] could not stamp MP4 dates", it) }
  }

  private fun insertPendingRow(options: SaveVideoInput): Uri {
    val values = ContentValues().apply {
      put(MediaStore.MediaColumns.DISPLAY_NAME, options.filename)
      put(MediaStore.MediaColumns.MIME_TYPE, MP4_MIME_TYPE)
      options.folder?.let { put(MediaStore.MediaColumns.RELATIVE_PATH, it) }
      putDates(options)
      put(MediaStore.MediaColumns.IS_PENDING, 1)
    }

    val collection = MediaStore.Video.Media.getContentUri(MediaStore.VOLUME_EXTERNAL_PRIMARY)
    return resolver.insert(collection, values)
      ?: throw SaveFailedException("The media store refused to create the file.")
  }

  private fun copyBytesInto(uri: Uri, sourcePath: String) {
    try {
      resolver.openOutputStream(uri)?.use { output ->
        File(sourcePath.removePrefix("file://")).inputStream().use { it.copyTo(output) }
      } ?: throw SaveFailedException("The media store gave no stream to write to.")
    } catch (error: Exception) {
      // A pending row with no bytes would linger invisibly forever.
      runCatching { resolver.delete(uri, null, null) }
      Log.w(TAG, "[save] write failed, row deleted", error)
      throw error
    }
  }

  /** Makes the row visible, re-sending the dates in the same update the scan is triggered by. */
  private fun publish(uri: Uri, options: SaveVideoInput) {
    writeValues(
      uri,
      ContentValues().apply {
        put(MediaStore.MediaColumns.IS_PENDING, 0)
        putDates(options)
      }
    )
  }

  /**
   * The last word on the dates: whatever the publish scan decided, these are the values §8 promised.
   *
   * The file's own timestamp is stamped too — DATE_MODIFIED is derived from it, so a media rescan
   * that ignores the column still lands on the right answer.
   */
  private fun assertDates(uri: Uri, options: SaveVideoInput) {
    if (options.capturedAtMs == null && options.modifiedAtMs == null) return

    writeValues(uri, ContentValues().apply { putDates(options) })
    options.modifiedAtMs?.let { stampUnderlyingFile(uri, it.toLong()) }
  }

  private fun ContentValues.putDates(options: SaveVideoInput) {
    options.capturedAtMs?.let { put(MediaStore.Video.Media.DATE_TAKEN, it.toLong()) }
    options.modifiedAtMs?.let {
      // DATE_MODIFIED is seconds, unlike DATE_TAKEN.
      put(MediaStore.MediaColumns.DATE_MODIFIED, it.toLong() / 1000)
    }
  }

  private fun writeValues(uri: Uri, values: ContentValues): ColumnWrite =
    runCatching { resolver.update(uri, values, null, null) }
      .fold(
        onSuccess = { ColumnWrite(rows = it, error = null) },
        onFailure = { ColumnWrite(rows = 0, error = it.message ?: it::class.java.simpleName) }
      )

  private fun describe(options: SaveVideoInput): String =
    "filename=${options.filename} folder=${options.folder} " +
      "capturedAt=${asDate(options.capturedAtMs?.toLong())} " +
      "modifiedAt=${asDate(options.modifiedAtMs?.toLong())}"

  /** What the media store actually holds for a row — the answer to "did the dates stick?". */
  private fun readColumns(uri: Uri): String {
    val takenMs = queryLong(uri, MediaStore.Video.Media.DATE_TAKEN)
    val modifiedSeconds = queryLong(uri, MediaStore.MediaColumns.DATE_MODIFIED)
    val addedSeconds = queryLong(uri, MediaStore.MediaColumns.DATE_ADDED)

    return "dateTaken=${asDate(takenMs)} " +
      "dateModified=${asDate(modifiedSeconds?.times(1000))} " +
      "dateAdded=${asDate(addedSeconds?.times(1000))} " +
      "relativePath=${queryString(uri, MediaStore.MediaColumns.RELATIVE_PATH)} " +
      "data=${queryString(uri, MediaStore.MediaColumns.DATA)}"
  }

  private fun asDate(epochMs: Long?): String =
    if (epochMs == null || epochMs <= 0) "none"
    else "${SimpleDateFormat("yyyy-MM-dd HH:mm:ss", Locale.US).format(Date(epochMs))} ($epochMs)"

  // MARK: - Metadata write-back

  /**
   * §8: carries the original capture and modified dates onto a newly saved asset.
   *
   * Every field is verified by reading it back, because a write the provider silently ignored looks
   * exactly like a successful one — and §8 asks for what could not be carried over, not for what we
   * attempted.
   *
   * Location cannot follow. `MediaStore.Video.Media.LATITUDE`/`LONGITUDE` were removed in Android
   * 10, and writing an ISO-6709 `©xyz` atom into the MP4 would need a muxer we do not have.
   */
  private fun applyAssetMetadata(
    assetId: String,
    metadata: AssetMetadataInput
  ): Map<String, Any> {
    val uri = Uri.parse(assetId)
    val capturedAtMs = metadata.capturedAtMs?.toLong()
    // DATE_TAKEN is milliseconds; DATE_MODIFIED is seconds.
    val modifiedAtSeconds = metadata.modifiedAtMs?.let { it.toLong() / 1000 }

    // One update for both columns: MediaProvider re-derives values on write, so a second update
    // can undo the first.
    val write = writeColumns(
      uri,
      buildMap<String, Long> {
        capturedAtMs?.let { put(MediaStore.Video.Media.DATE_TAKEN, it) }
        modifiedAtSeconds?.let { put(MediaStore.MediaColumns.DATE_MODIFIED, it) }
      }
    )

    val applied = mutableListOf<String>()
    val skipped = mutableListOf<Map<String, String>>()

    capturedAtMs?.let { expected ->
      // DATE_TAKEN is ours to set, so it has to read back exactly.
      val actual = queryLong(uri, MediaStore.Video.Media.DATE_TAKEN)
      if (write.succeeded && actual == expected) applied.add("capturedAt")
      else skipped.add(field("capturedAt", write.explain("wrote $expected, read back $actual")))
    }

    modifiedAtSeconds?.let { expected ->
      // The provider owns DATE_MODIFIED and recomputes it from the file, so an exact read-back is
      // the wrong success test — stamp the file too and accept the write.
      val stamped = metadata.modifiedAtMs?.let { stampUnderlyingFile(uri, it.toLong()) } ?: false
      if (write.succeeded || stamped) applied.add("modifiedAt")
      else skipped.add(field("modifiedAt", write.explain("wrote $expected, and the file stamp failed")))
    }

    if (metadata.latitude != null && metadata.longitude != null) {
      skipped.add(
        field(
          "location",
          "Android removed the media store's location columns in Android 10."
        )
      )
    }

    return mapOf("applied" to applied, "skipped" to skipped)
  }

  /** What a write attempt did, kept whole so a skip can say why rather than just that it failed. */
  private data class ColumnWrite(val rows: Int, val error: String?) {
    val succeeded: Boolean get() = rows > 0

    fun explain(detail: String): String = when {
      error != null -> "The media store threw: $error"
      rows == 0 -> "The media store accepted no rows for this asset."
      else -> "The media store overwrote the value: $detail."
    }
  }

  private fun writeColumns(uri: Uri, columns: Map<String, Long>): ColumnWrite {
    if (columns.isEmpty()) return ColumnWrite(rows = 0, error = null)

    return writeValues(
      uri,
      ContentValues().apply { columns.forEach { (name, value) -> put(name, value) } }
    )
  }

  private fun stampUnderlyingFile(uri: Uri, modifiedAtMs: Long): Boolean {
    val path = queryString(uri, MediaStore.MediaColumns.DATA) ?: return false
    return runCatching { File(path).setLastModified(modifiedAtMs) }.getOrDefault(false)
  }

  private fun field(name: String, reason: String) = mapOf("field" to name, "reason" to reason)

  // MARK: - Foreground service

  /** Counts start requests, so a queued stop can tell when it has been superseded by a new session. */
  @Volatile
  private var serviceStartCount = 0

  /** Starting is only ever called while the app is on screen, where a service start is allowed. */
  private fun startService(options: ServiceNotification) {
    serviceStartCount += 1
    CompressionForegroundService.stopRequested = false

    val live = CompressionForegroundService.running
    if (live != null && !live.finishing) {
      // The service survived the previous session. A second start intent would open a fresh
      // startForeground obligation that a racing stop could leave unmet — and an unmet obligation
      // kills the app. Re-titling the live notification is all a new session needs.
      updateNotification(options)
      return
    }

    val intent = Intent(context, CompressionForegroundService::class.java).apply {
      action = CompressionForegroundService.ACTION_START
      putExtra(CompressionForegroundService.EXTRA_TITLE, options.title)
      putExtra(CompressionForegroundService.EXTRA_PROGRESS, options.progress)
      putExtra(CompressionForegroundService.EXTRA_ELAPSED, options.elapsed)
      putExtra(CompressionForegroundService.EXTRA_REMAINING, options.remaining)
    }

    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
      context.startForegroundService(intent)
    } else {
      context.startService(intent)
    }
  }

  /**
   * Progress goes straight to the notification rather than through the service.
   *
   * Android 12+ rejects `startForegroundService` from the background — which is exactly where a
   * compression spends its time — so an intent per update stopped the notification from moving as
   * soon as the app left the screen. `notify()` on the service's own id has no such restriction and
   * updates the same notification in place.
   */
  private fun updateNotification(options: ServiceNotification) {
    CompressionNotification.post(
      context,
      CompressionStatus(options.title, options.progress, options.elapsed, options.remaining)
    )
  }

  /**
   * Ends the service through the live instance instead of another start intent, which the platform
   * would refuse from the background — the case where jobs actually finish.
   *
   * Every service call returns `Unit` on purpose: `startService` answers with a `ComponentName`,
   * and an `AsyncFunction` whose body ends in one tries to send that across the bridge and rejects.
   */
  private fun stopService() {
    val requestedAt = serviceStartCount
    val appContext = context.applicationContext

    Handler(Looper.getMainLooper()).post {
      // A newer session claimed the service while this stop sat in the queue — leave it alone.
      if (serviceStartCount != requestedAt) return@post

      // The start intent may not have been delivered yet; the flag makes the service stop itself
      // right after it satisfies the startForeground obligation, instead of the stop being lost.
      CompressionForegroundService.stopRequested = true
      CompressionForegroundService.running?.finish()
      // The service may never have started — or already have died — but the notification is ours.
      CompressionNotification.cancel(appContext)
    }
  }
}

class SaveVideoInput : Record {
  @Field val path: String = ""

  @Field val filename: String = "video.mp4"

  /** MediaStore `RELATIVE_PATH`, e.g. "DCIM/Camera/". Null saves to the default location. */
  @Field val folder: String? = null

  @Field val capturedAtMs: Double? = null

  @Field val modifiedAtMs: Double? = null
}

class SaveFailedException(message: String) : Exception(message)

class ServiceNotification : Record {
  @Field val title: String = "Compressing video"

  @Field val progress: Int = 0

  /** e.g. "1 min 12 s elapsed". */
  @Field val elapsed: String = ""

  /** e.g. "2 min 5 s left". */
  @Field val remaining: String = ""
}

class AssetMetadataInput : Record {
  @Field val capturedAtMs: Double? = null

  @Field val modifiedAtMs: Double? = null

  @Field val latitude: Double? = null

  @Field val longitude: Double? = null
}

private fun parseIso6709(value: String?): Map<String, Double>? {
  if (value.isNullOrBlank()) return null

  // e.g. "+52.3676+004.9041/" — sign-prefixed latitude then longitude.
  val match = Regex("([+-]\\d+\\.?\\d*)([+-]\\d+\\.?\\d*)").find(value) ?: return null
  val latitude = match.groupValues[1].toDoubleOrNull() ?: return null
  val longitude = match.groupValues[2].toDoubleOrNull() ?: return null

  return mapOf("latitude" to latitude, "longitude" to longitude)
}
